Fix division by zero and array out of bounds in JD acos()-LUT

XDA-Bam opened the pull request:
### Description

The current LUT implementation has two problems:
- Input values from `1-1/2^14 = 0.99994` to `1-1/2^15 = 0.99997` (absoulte) return an `acos()` of zero, causing a division by zero  in https://github.com/MarlinFirmware/Marlin/blob/e380498512a325fe1e70b20d4c60a81f606eddbf/Marlin/src/module/planner.cpp#L2387
- Input values from `1-1/2^15` to `1` cause an array out of bounds error (see [this post](https://github.com/MarlinFirmware/Marlin/issues/17920#issuecomment-636395133) for details).

This PR fixes both of these problems. Also fixes a couple of typos in the comments. The new LUT never reaches zero inside the `cos()`-interval of [-0.999999, 0.999999] used in the planner.

New array values are computed in MATLAB using single precision floating point, not in C++. There seem to be small differences, starting at the 6th to 7th digit. If you're not comfortable with that, let me know (or be so kind to run it on your Arduino and post the results).

### Benefits

Avoids division by zero or NaN. Should reduce stutter resulting from zero junction speeds at angles close to 180° when using the `acos()`-LUT.

### Related Issues

Users in #17920 and #17342 reported less stuttering when _not_ using the LUT for `acos()` (i.e. running with `//#define JD_USE_LOOKUP_TABLE`). This should reduce stuttering when running the LUT.

XDA-Bam committed:
Fix division by zero in JD acos()-LUT

The LUT had two problems:
- Input values from 1-1/2^14 (0.99994) to 1-1/2^15 (0.99997) would return an acos of zero, causing a divide by zero later on.
- Input values from 1-1/2^15 (0.99997) to 1 would cause an array out of bounds error.

This fixes both of them.

thinkyhead committed:
Update planner.cpp

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bf2_fix_jd_math_18175

XDA-Bam committed:
Fix typo

For `jd_lut_count = 16`, the result should be 2^(16-1), not (2^16)-1.

XDA-Bam commented:
@thinkyhead I changed something right before pushing the branch and slipped in a typo: It should not be
```C++
static constexpr uint16_t jd_lut_tll   = 1 << jd_lut_count -1;
```
but
```C++
static constexpr uint16_t jd_lut_tll   = 1 << (jd_lut_count -1);
```

With your change, the line becomes
```C++
static constexpr uint16_t jd_lut_tll   = _BV(jd_lut_count - 1) ;
```

AnHardt commented:
What matters is the position of the first set bit (from the left).
That is constant for all of the versions.

1 << 7 = 0b1000'0000
(1 << 7) - 1 = 0b0111'1111
1 << (7-1) = 0b0100'0000

AnHardt commented:
Sorry. It does matter.
Here `__builtin_clz(uint16_t((1.0f - t) * jd_lut_tll)) - jd_lut_tll0;` if makes nearly factor two (or +-1 after clz() ) in `(1.0f - t) * jd_lut_tll)`.

XDA-Bam commented:
Yes. `idx` is gonna be `= -1` for inputs below 0.5f, if you use `jd_lut_tll   = 1 << jd_lut_count -1`.

There is, however, another potential pitfall: For values of `junction_cos_theta` very, very, very close to zero, this might still return `idx = -1`. At least it does so in my MATLAB version for inputs of 0.00000002f and below. I will see if I can test this in Marlin later on. Would be easy to fix, though by going
```c++
const int16_t idx = (t <= 0.00000003f) ? 0 : __builtin_clz(uint16_t((1.0f - t) * jd_lut_tll)) - jd_lut_tll0;
```

XDA-Bam committed:
Catch second array out of bounds error in JD acos()-LUT

For very, very small values of `t` below 0.00000003f, `idx` could become -1. Problem caused by `(1.0f - t) * jd_lut_tll` rounding to 2^15, which has 0 leading zeros as uint16_t.

XDA-Bam commented:
Indeed it returned `idx=-1` for values below 0.00000003f. Fixed now. Damn you, numerics! 👊

XDA-Bam commented:
bugfix-2.0.x @ a1f3d2f:
![JD bugfix-2 0 x-a1f3d2f](https://user-images.githubusercontent.com/1209896/83578277-a0a1d980-a536-11ea-86d0-56c900681543.png)

bugfix-2.0.x @ a1f3d2f + this PR:
![JD bugfix-2 0 x-a1f3d2f + #18175_crop](https://user-images.githubusercontent.com/1209896/83578300-aa2b4180-a536-11ea-83ec-ba55466d401f.png)

Ender 3 Pro, stock board, config close to default ([Config.zip](https://github.com/MarlinFirmware/Marlin/files/4720121/Config.zip)), [STL](https://github.com/MarlinFirmware/Marlin/issues/17920#issuecomment-630900799).

thinkyhead commented on Marlin/src/module/planner.cpp:
```diff
@@ -2341,22 +2340,24 @@ bool Planner::_populate_block(block_t * const block, bool split_move,
-            static constexpr int16_t  jd_lut_tll0  = __builtin_clz(jd_lut_tll) + 1; // i.e., 16 - jd_lut_count
+            static constexpr int16_t  jd_lut_count = 16;
+            static constexpr uint16_t jd_lut_tll   = _BV(jd_lut_count - 1);
+            static constexpr int16_t  jd_lut_tll0  = __builtin_clz(jd_lut_tll) + 1; // i.e., 16 - jd_lut_count + 1
```
Of course, all this rigamarole can be skipped since we know that `__builtin_clz(_BV(n))` will always give the result `15 - n`. We are starting from the point of knowing the bit index, so counting the leading zeros is … indulgent.

thinkyhead commented:
Lovely…. Time to merge!

qwewer0 commented:
Nice job :thumbsup:
**Left**: bugfix before this - **Right**: with this fix
![IMG_20200603_080532](https://user-images.githubusercontent.com/57561110/83601798-95b86a80-a571-11ea-98ca-313210eb5e2a.jpg)
Sliced with Cura.
[Gcode.zip](https://github.com/MarlinFirmware/Marlin/files/4721313/Gcode.zip)
